### PR TITLE
React feature flags

### PR DESCRIPTION
**Pull request by elsassph**
- Adding `-D react_ver=major.minor` to control feature flags (defaults to latest)
- Add React 16.9 unsafe lifecycle methods
- Retrofit flags to previous features that were just enabled

### CODE REVIEW

File: test/src/react/ReactComponent.hx [hunk 1]
```diff
@@ -1,5 +1,10 @@
 package react;
+
+#if (haxe_ver >= 4)
```
Comment (kLabz) [inline test/src/react/ReactComponent.hx]: `#if haxe4`? (I don't think haxe 4 previews still need to be supported)

File: src/lib/react/ReactTypeMacro.hx [hunk 2]
```diff
@@ -12,6 +12,38 @@ class ReactTypeMacro
 	static public inline var ENSURE_RENDER_OVERRIDE_BUILDER = 'EnsureRenderOverride';
 
 	#if macro
+
+	// define React feature flags based on -D haxe_ver (default to latest)
+	public static function setFlags() {
+		var ver = Context.defined("react_ver") ? Context.definedValue("react_ver") : "16.12";
+		var match = ~/([0-9]+).([0-9]+)/;
+		if (!match.match(ver)) {
+			Context.fatalError("Invalid `haxe_ver` specified: " + Context.definedValue("react_ver"), Context.currentPos());
```
Comment (Gama11) [inline src/lib/react/ReactTypeMacro.hx]: `haxe_ver`? :)

File: src/lib/react/ReactTypeMacro.hx [hunk 3]
```diff
@@ -12,6 +12,38 @@ class ReactTypeMacro
 	static public inline var ENSURE_RENDER_OVERRIDE_BUILDER = 'EnsureRenderOverride';
 
 	#if macro
+
+	// define React feature flags based on -D haxe_ver (default to latest)
```
Comment (sergeykgh) [inline src/lib/react/ReactTypeMacro.hx]: based on -D react_ver .. ?

File: src/lib/react/ReactTypeMacro.hx (hunk 2 above)
Comment (elsassph) [reply src/lib/react/ReactTypeMacro.hx]: Hum

File: src/lib/react/React.hx [hunk 4]
```diff
@@ -51,7 +55,9 @@ extern class React
 		If you are using an earlier release of React, use callback refs instead
 		https://reactjs.org/docs/refs-and-the-dom.html#callback-refs
 	**/
+	#if react_ref_api
```
Comment (Gama11) [inline src/lib/react/React.hx]: I think Haxe uses the first doc comment it comes across, so if you don't include the doc comment in the conditionals, the following field will have the wrong one.

Comment (elsassph) [reply src/lib/react/React.hx]: Thanks for the tip

File: src/lib/react/ReactTypeMacro.hx [hunk 5]
```diff
@@ -12,6 +12,39 @@ class ReactTypeMacro
...
+			define("react_context_api");
+			define("react_snapshot_api");
+		}
+		if (semver(version, [16, 9])) {
+			define("react_unsafe_lifecycle");
+		}
+	}
+
+	static function semver(target: Array<Int>, required: Array<Int>) {
+		if (target[0] > required[0]) return true;
```
Comment (sergeykgh) [inline src/lib/react/ReactTypeMacro.hx]: if (target[0] != required[0]) return target[0] > required[0];

Comment (elsassph) [reply src/lib/react/ReactTypeMacro.hx]: hmm indeed there's a bug here

**Review by sergeykgh — APPROVED** [review]

**Review by kLabz — APPROVED** [review]

**Pull request merged**